### PR TITLE
Support/Handle exception for rate limits.

### LINKED ISSUE #6: Support/Handle exception for rate limits.
In its current state, the bot tests whether it can login to Twitter and does **not handle** or **sufficiently log** any sort of rate limits imposed by Twitter.

Of course, reaching a rate limit can be hard for an account that just tweets once every few days, but a few external factors may also come into play, impose limitations on the bot account and then crash the bot itself.

That alone resulted in an outage that lasted for a few days, after my personal account (as well as all other accounts that are owned by me) got locked out, due to [Twitter's mitigation efforts regarding COVID-19 misinformation](https://blog.twitter.com/en_us/topics/company/2020/An-update-on-our-continuity-strategy-during-COVID-19.html).

The exception that's normally used by tweepy is [tweepy.RateLimitError](http://docs.tweepy.org/en/latest/api.html#RateLimitError). But since the exception inherits from [tweepy.TweepError], that results in the bot mishandling the situation and treating a rate limit as an "authentication error".

The bot should check for rate limits FIRST when an exception arises and **overlook** the error, without interrupting the execution of the bot itself.

https://github.com/AlwaysLivid/TinyFlowerBeds/blob/9b571a32425c854c4cab8093932da1ca0395f679/src/bot.py#L118

### PR DESCRIPTION
**Pull request by Shankhanil**
Two changes have been made wrt Issue #6 

1. The bot checks for 2 errors only: 1. Authentication error caused by wrong tokens, and 2. RateLimitError caused by exceeding rate limits in twitter. Any other exception caused by external factors will be ignored.
2. Cooldown time is in seconds not minutes. 

I really need a test suite to test whether the update works properly or not. @AlwaysLivid

### CODE REVIEW

File: src/bot.py [hunk 1]
```diff
@@ -1,26 +1,31 @@
...
 from config import *
 from random import randint
-import sys
-import credentials
+
 print("""
                        TinyFlowerBeds
-
             Copyright (C) 2018-2019 AlwaysLivid
-
```
Comment (AlwaysLivid) [inline src/bot.py:19]: Why were those spaces removed?

Comment (n0toose) [inline src/bot.py]: Why were those spaces removed? That's irrelevant to the pull request, the issue and it's purpose. Please open new issues before making changes like this.

File: src/bot.py [hunk 2]
```diff
@@ -1,26 +1,31 @@
...
 =============================================================
 ======================= DISCLAIMER ==========================
 =============================================================
 This program comes with ABSOLUTELY NO WARRANTY.
 This is free software, and you are welcome to redistribute it
 under certain conditions; read the LICENSE file for details.
 =============================================================
-
```
Comment (n0toose) [inline src/bot.py]: Why was this removed? I want to keep a newline before the actual logging of the bot.

File: src/bot.py [hunk 3]
```diff
@@ -1,26 +1,31 @@
...
 """)
 
+FLAG = True
```
Comment (n0toose) [inline src/bot.py]: What's the point of the flag? It's not explained and not used practically anywhere.

File: src/bot.py [hunk 4]
```diff
@@ -34,10 +39,53 @@
 maxinterval = config['maxinterval']
 
 cooldown = randint(mininterval, maxinterval) * 24 * 60 * 60
+
```
Comment (n0toose) [inline src/bot.py]: Why did you add a newline above the comment which explained the purpose of the `cooldown` variable?

File: src/bot.py [hunk 5]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
                 else:
                     api.update_status(self.generate_batch())
                     logging.info("Tweeted out a new flower bed!")
-                    logging.info("The next tweet is scheduled to be made in {} minutes".format(cooldown))
+                    logging.info("The next tweet is scheduled to be made in {} seconds".format(cooldown))
```
Comment (n0toose) [inline src/bot.py]: This is a valid mistake that I'd be happy to address-- in another issue and pull request.

File: src/bot.py [hunk 6]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
...
                     api.update_status(self.generate_batch())
                     logging.info("Tweeted out a new flower bed!")
-                    logging.info("The next tweet is scheduled to be made in {} minutes".format(cooldown))
+                    logging.info("The next tweet is scheduled to be made in {} seconds".format(cooldown))
                     time.sleep(cooldown)
-            except:
-                print(sys.exc_info())
-                logging.critical("Tweeting failed due to ratelimit. Waiting {} more minutes.".format(cooldown))
-                time.sleep(cooldown)
+                    raise tweepy.TweepError("error")
```
Comment (n0toose) [inline src/bot.py]: Why are you raising an error? The syntax isn't right here, you'd want to type it like this;
```
time.sleep(cooldown)
raise tweepy.TweepError("error")
```

But this just raises an error for literally no reason.

File: src/bot.py [hunk 7]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
                 else:
                     api.update_status(self.generate_batch())
                     logging.info("Tweeted out a new flower bed!")
-                    logging.info("The next tweet is scheduled to be made in {} minutes".format(cooldown))
+                    logging.info("The next tweet is scheduled to be made in {} seconds".format(cooldown))
                     time.sleep(cooldown)
-            except:
```
Comment (n0toose) [inline src/bot.py]: It would be great if we could handle **ratelimits** only and retaining the functionality instead of completely crashing the entire bot.

File: src/bot.py [hunk 8]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
...
-                print(sys.exc_info())
-                logging.critical("Tweeting failed due to ratelimit. Waiting {} more minutes.".format(cooldown))
-                time.sleep(cooldown)
+                    raise tweepy.TweepError("error")
+            except Exception as ex:
+                exception = type(ex).__name__
                 
+                if exception == 'RateLimitError':
+                    logging.critical("Tweeting failed due to ratelimit. Waiting {} more seconds.".format(cooldown))
+                    time.sleep(cooldown)
```
Comment (n0toose) [inline src/bot.py:130]: Could you keep the exception handlers in the same place as the other exception handlers? (Bottom of the script)

File: src/bot.py [hunk 9]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
...
+                else:
+                    logging.critical("TweepError not RateLimitError.")
```
Comment (n0toose) [inline src/bot.py]: Then what? Shouldn't the entire bot just crash at this point? That change has no effect.

File: src/bot.py [hunk 10]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
...
-
+        user = api.me()
+        self.user_info(api, user)
+        logging.info("Successfully logged in!")
+        self.tweet_loop(api)
 
 if __name__ == '__main__':
     try:
-        
+        FLAG = True
```
Comment (n0toose) [inline src/bot.py]: What's the point of this boolean?

File: src/bot.py [hunk 11]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
...
         TwitterBot = Bot()
-
```
Comment (n0toose) [inline src/bot.py]: How are the line deletions relevant to the issue? Again, this is abandoned code, I'd be glad to discuss this, but why do that here?

File: src/bot.py [hunk 12]
```diff
@@ -75,40 +123,43 @@ def tweet_loop(self, api):
...
+        FLAG = True
         TwitterBot = Bot()
-        
         TwitterBot.main()
-    except:
-      print(sys.exc_info())
-      quit()
+        
+    except Exception as ex:
+        exception = type(ex).__name__
```
Comment (n0toose) [inline src/bot.py]: This isn't the Python way of doing this.

Something like [this](https://stackoverflow.com/questions/6095717/python-one-try-multiple-except/6095782) would look way better.

```python
try:
    something()
except SpecificExceptionClass:
    handling1()
except Exception as e:
    handling2()
```

**Review by n0toose — COMMENTED** [review]
Hello, please explain your changes, keep irrelevant changes separate and please address one issue per pull request.

This is doing quite well so far.

**Comment by n0toose** [comment]
>     1. The bot checks for 2 errors only: 1. Authentication error caused by wrong tokens, and 2. RateLimitError caused by exceeding rate limits in twitter. Any other exception caused by external factors will be ignored.

That's some good thinking, but I firmly believe that no exception should be ignored. If the program can't handle it, it should crash, after doing what it should reasonably handle. That's what crashes are there for.

There

>     2. Cooldown time is in seconds not minutes.

**This is a valid pedantic issue,** however, it is not currently relevant to the changes you're proposing and the issue which you want to solve.
 
> I really need a test suite to test whether the update works properly or not. @AlwaysLivid

I'd advise you to open a new Twitter account and run the bot there. I can't see how a test suite would help here, or what it would possibly do.

**Please do not open another pull request**, instead, [squash](https://github.com/wprig/wprig/wiki/How-to-squash-commits) the commits and reapply your changes. Only keep one of them in the pull request, and open another one that's meant to deal with the other issue. Stay on-topic.

File: src/bot.py (hunk 3 above)
Comment (Shankhanil) [reply src/bot.py]: Apologies. I was experimenting with exception, nothing to do with actual code.

File: src/bot.py (hunk 5 above)
Comment (Shankhanil) [reply src/bot.py]: done. I will do it

File: src/bot.py (hunk 6 above)
Comment (Shankhanil) [reply src/bot.py]: Apologies. I was experimenting with exception, nothing to do with actual code.

File: src/bot.py (hunk 8 above)
Comment (Shankhanil) [reply src/bot.py:130]: suggestion taken.

File: src/bot.py (hunk 10 above)
Comment (Shankhanil) [reply src/bot.py]: Apologies. I was experimenting with exception, nothing to do with actual code. I will update in the upcoming pull request

File: src/bot.py (hunk 12 above)
Comment (Shankhanil) [reply src/bot.py]: Pls explain what is wrong with the syntax. 

`except Exception as e:
       handling()`

File: src/bot.py (hunk 5 above)
Comment (n0toose) [reply src/bot.py]: > done. I will do it

I'd appreciate it if we could address this after this pull request, alongside other pedantic changes as well.

File: src/bot.py (hunk 12 above)
Comment (n0toose) [reply src/bot.py]: There's nothing wrong with it, but "Exception" should be used as a fallback, after all the exceptions you can handle.

Using "Exception" and then if... else cases is definitely not clean.

File: src/bot.py (hunk 9 above)
Comment (Shankhanil) [reply src/bot.py]: SO i will crash the bot if it's not an authentication error or an rate limit error. Suggestion taken

File: src/bot.py (hunk 12 above)
Comment (Shankhanil) [reply src/bot.py]: I am catching all forms of exceptions, and then I am using If as a mode to check the type of exception. 
it's basically, 
if ex is Tweeperror -------> authentication error
if ex is RateLimitError -------> rate limit error
else --------> other error, crash

Comment (n0toose) [reply src/bot.py]: > I am catching all forms of exceptions, and then I am using If as a mode to check the type of exception.
> it's basically,
> if ex is Tweeperror -------> authentication error
> if ex is RateLimitError -------> rate limit error
> else --------> other error, crash

Yeah, but that's something that you'd probably do in another language, but catching the Exception, storing it as a string and then comparing it in multiple places across the codebase is not the recommended way to do it. The main process is running under the main() function-- if it crashes, ***ALL*** sorts of exceptions will have to be handled in the bottom of the package, and not be split for no reason. It makes the code dirtier.

`try... except` blocks in Python are "self-sufficient" in a sense. "Try this, and if this happens, then do that". What you're proposing is something along the lines of "Try this, unless if something happens. Assign what happened in this variable. If that something is this, then do this. If that something is that, then do that.". So, why should we create more blocks? There's already a design that works and is mostly universal across Python packages and programs, so why would we not do that as well? It's essentially a standard.

File: src/bot.py [hunk 13]
```diff
@@ -123,23 +120,16 @@ def tweet_loop(self, api):
...
-                if exception == 'RateLimitError':
-                    logging.critical("Tweeting failed due to ratelimit. Waiting {} more seconds.".format(cooldown))
-                    time.sleep(cooldown)
-                else:
-                    logging.critical("TweepError not RateLimitError.")
-                    
+            except tweepy.RateLimitError:
+                logging.critical("Tweeting failed due to ratelimit. Waiting {} more minutes.".format(cooldown))
+                time.sleep(cooldown)
+
```
Comment (AlwaysLivid) [inline src/bot.py]: There are two handlers for tweepy.RateLimitError

Comment (n0toose) [inline src/bot.py:132]: There are two handlers for tweepy.RateLimitError

File: src/bot.py [hunk 14]
```diff
@@ -151,15 +141,13 @@ def main(self):
...
-            time.sleep(10)
-            quit()
\ No newline at end of file
+    except tweepy.RateLimitError:
+        logging.critical("Tweeting failed due to ratelimit. Waiting {} more minutes.".format(cooldown))
+        time.sleep(cooldown)
+    except tweepy.TweepError:
+      logging.critical("Authentication Error!")
+      logging.info("Please validate your credentials.")
+      time.sleep(10)
```
Comment (n0toose) [inline src/bot.py]: This is starting to look WAY better, although I'd remove `time.sleep()`

**Review by n0toose — COMMENTED** [review]
This looks great, although I'll note that there are two issues that haven't been addressed regarding removed new lines. Great job!

File: src/bot.py (hunk 13 above)
Comment (Shankhanil) [reply src/bot.py:132]: I was experimenting with the code when I mistakenly removed those lines. I will try and keep un-related code untouched. Cheers!

File: src/bot.py (hunk 14 above)
Comment (Shankhanil) [reply src/bot.py]: Noted

**Review by Shankhanil — COMMENTED** [review]
If you find any removed space, or comment, it might be because I was tinkering with the  code to run some tests and experiments. Apologies for those.

**Review by Shankhanil — COMMENTED** [review]
I have taken your suggestion, and made the changes. Pls compare the latest commit to your original bot.py file. @AlwaysLivid

**Comment by n0toose** [comment]
> I have taken your suggestion, and made the changes. Pls compare the latest commit to your original bot.py file. @AlwaysLivid

Looks good to me! I'll make sure to test it later before merging it and I'll let you know if something's wrong again. Thanks!

File: src/bot.py [hunk 15]
```diff
@@ -125,9 +125,6 @@ def tweet_loop(self, api):
                     logging.info("Tweeted out a new flower bed!")
                     logging.info("The next tweet is scheduled to be made in {} minutes".format(cooldown))
                     time.sleep(cooldown)
-            except tweepy.RateLimitError:
-                logging.critical("Tweeting failed due to ratelimit. Waiting {} more minutes.".format(cooldown))
-                time.sleep(cooldown)
```
Comment (n0toose) [inline src/bot.py:130]: Why did you remove both handlers? There should be at least one that halts the bot before it carries on.

File: src/bot.py [hunk 16]
```diff
@@ -146,8 +143,17 @@ def main(self):
     try:
         TwitterBot = Bot()
         TwitterBot.main()
+        
+        
+
```
Comment (n0toose) [inline src/bot.py]: Why so many spaces?

File: src/bot.py [hunk 17]
```diff
@@ -146,8 +143,17 @@ def main(self):
     try:
         TwitterBot = Bot()
         TwitterBot.main()
+        
+        
+        
+    except tweepy.RateLimitError:
+        logging.critical("Tweeting failed due to ratelimit. Waiting {} more minutes.".format(cooldown))
+        time.sleep(cooldown)
```
Comment (n0toose) [inline src/bot.py]: Please review the logic here. Nothing will happen after the `time.sleep(cooldown)` function.

Maybe it would be best to move both handlers at the function which handles the tweets for now.

Comment (Shankhanil) [reply src/bot.py]: That is what I came up, after I tested the bot. We should keep the authentication error handler at the main function, and the RateLimit and generic Exception handler at the bot.

File: src/bot.py (hunk 16 above)
Comment (Shankhanil) [reply src/bot.py]: I will remove them.

File: src/bot.py (hunk 15 above)
Comment (Shankhanil) [reply src/bot.py:130]: ok. So the main will contain only the TweepError exception, and the bot will contain RateLimit error and a generic exception handler.

File: src/bot.py (hunk 17 above)
Comment (Shankhanil) [reply src/bot.py]: The above approach is not working properly. I am keeping the generic exception handler at the main function to avoid ambiguity.

**Comment by Shankhanil** [comment]
Made the changes suggested previously:
1. Removed extra spaces mentioned previously/
2. Kept the RateLimit error handler in the bot, while the TweepError handler and the generic exception handler in the main driver code.

Have a look @AlwaysLivid

**Review by Shankhanil — COMMENTED** [review]
Perfect changes @AlwaysLivid . Is it ready to be merged?

**Comment by n0toose** [comment]
> Perfect changes @AlwaysLivid . Is it ready to be merged?

Looks absolutely good to me-- I'll have to verify that, however.

**Review by Shankhanil — COMMENTED** [review]
If you remove that, how till it catch any exception that is neither authentication error nor RateLimit error?

**Comment by n0toose** [comment]
> If you remove that, how till it catch any exception that is neither authentication error nor RateLimit error?

In standard Python code, you're not supposed to do that. I looked into it a bit more-- doing that just removes details about the error, but I'll add again soon after I make a small change based on your own code (and you'll be given proper attribution).
